environ.go

```diff
@@ -0,0 +1,304 @@
+		} else {
+			for _, mKey := range keys(*m) {
+				if keeperKey == mKey {
+					keeping[mKey] = (*m)[mKey]
+					found = true
+
+					break
+				}
+			}
+		}
+
+		if !found {
+			missing = append(missing, keeperKey)
+		}
+	}
+
+	*m = keeping
+
+	sort.Strings(missing)
+
+	return missing
+}
+
+// Drop scans the Environ looking for keys matching the droppers slice and
+// drops them while keeping all others.
+//
+// It returns the slice of keys it could not find.
+//
+// If a key in the droppers list has the suffix "*", it will
+// use wild-card logic to capture columns.
+func (e *Environ) Drop(droppers ...string) (missing []string) {
+	m := e.AsMap()
+	missing = drop(m, droppers)
+
+	defer e.writeLocker()()
+
+	e.m = m
+
+	return missing
+}
+
+func drop(m map[string]string, droppers []string) (missing []string) {
+	mm := make(map[string]struct{}, len(droppers))
+	for _, dropKey := range droppers {
+		var found bool
+		if strings.HasSuffix(dropKey, "*") {
+			kprefix := strings.TrimSuffix(dropKey, "*")
+			// hold a streak state to stop when we hit the last matching key in the
+			// map, since the keys are sorted.
+			var streak bool
+			for _, mKey := range keys(m) {
+				if strings.HasPrefix(mKey, kprefix) {
+					delete(m, mKey)
+					streak = true
+					// we don't break in this case, as we may
+					// have multiple matches.
+				} else if streak {
+					found = true
+
+					break
+				}
+			}
+		} else {
+			if _, ok := m[dropKey]; ok {
+				delete(m, dropKey)
+				found = true
+			}
+		}
+
+		if !found {
+			mm[dropKey] = struct{}{}
+		}
+	}
+
+	missing = make([]string, 0, len(missing))
+	for k := range mm {
+		missing = append(missing, k)
+	}
+
+	// Consistently report keys
+	sort.Strings(missing)
+
+	return missing
+}
+
+// Keys returns the map's keys in lexical order.
+func (e *Environ) Keys() []string {
+	defer e.readLocker()()
+
+	return keys(e.m)
+}
+
+func keys(m map[string]string) []string {
+	var keys []string
+	for k := range m {
+		keys = append(keys, k)
+	}
+
+	// Consistently report keys
+	sort.Strings(keys)
+
+	return keys
+}
+
+// AsMap returns a copy of the internal map structre of the Environ.
+func (e *Environ) AsMap() map[string]string {
+	defer e.readLocker()()
+
+	return copyMap(e.m)
+}
+
+func copyMap(e map[string]string) map[string]string {
+	res := make(map[string]string, len(e))
+	for k, v := range e {
+		res[k] = v
+	}
+
+	return res
+}
+
+func envSliceAsMap(env []string) map[string]string {
+	m := make(map[string]string, len(env))
+	for _, v := range env {
+		// in case we're reading a .env file with comments or blank lines
+		if strings.HasPrefix(v, "#") || v == "" {
+			continue
+		}
+		if !strings.Contains(v, "=") {
+			continue
+		}
+		kv := strings.SplitN(v, "=", 2)
+		if len(kv) == 2 {
+			m[kv[0]] = kv[1]
+
+			continue
+		}
+	}
+
+	return m
+}
+
+// AsSlice emits the contents of the Environ as a slice of string
+// with a "key=value" format.
+func (e *Environ) AsSlice() []string {
+	defer e.readLocker()()
```

Comment:
i'm trying to wrap my head around how this is working. Is this immediately invoking e.readLocker() which then returns a function, but that function is deferred to execute. What is the reason for this over an explicit like

```
e.Lock()
defer e.Unlock()
....
```

Reply:
The reason is often forgetting to defer the unlock, mostly. I have a test that would work for either style, and I can replace this "clever" defer with a less clever inlined pair in all locations.

Reply:
this also prevents:

```
e.Lock()
defer e.RUnlock()
```

Reply:
In fact, that was the problem I was trying to solve with that construct, because I forgot several of the defers or used the wrong unlock function before implementing the locker closure.